File: src/main/kotlin/github/io/vocabmate/domain/vocabs/Vocab.kt
package github.io.vocabmate.domain.vocabs

import com.faunadb.client.types.FaunaConstructor
import com.faunadb.client.types.FaunaField
import io.micronaut.core.annotation.Introspected
import java.time.Instant

@Introspected
data class Vocab(
    val id: String? = null,
    val word: String,
    val partOfSpeech: PartOfSpeech,
    val definition: String,
    val examples: List<String> = emptyList(),
    val synonyms: List<String> = emptyList(),
    val antonyms: List<String> = emptyList(),
    val lastUpdated: Instant? = null,
) {

    companion object {
        @JvmStatic
        @Suppress("unused")
        @FaunaConstructor
        fun fromFauna(
            @FaunaField("word") word: String,
            @FaunaField("partOfSpeech") partOfSpeech: String,
            @FaunaField("definition") definition: String,
            @FaunaField("examples") examples: List<String>?,
            @FaunaField("synonyms") synonyms: List<String>?,
            @FaunaField("antonyms") antonyms: List<String>?,
        ) = Vocab(
            word = word,
            partOfSpeech = PartOfSpeech.valueOf(partOfSpeech),
            definition = definition,
            examples = examples ?: emptyList(),
            synonyms = synonyms ?: emptyList(),
            antonyms = antonyms ?: emptyList(),
        )
    }

    @Introspected
    enum class PartOfSpeech(val shortForm: String) {
        Noun("n"), Verb("v"), Adjective("adj"), Adverb("adv");

        companion object {
            @JvmStatic
            fun fromFullWord(value: String) = values().find {
                it.name.equals(value, ignoreCase = true)
            }
        }
    }
}

File: v2/vocab-mate/src/vocab.rs
use chrono::prelude::*;
use serde::Serialize;

#[derive(Debug, Serialize)]
pub struct Vocab {
    pub id: String,
    pub word: String,
    pub(crate) part_of_speech: PartOfSpeech,
    pub definition: String,
    pub examples: Vec<String>,
    pub synonyms: Vec<String>,
    pub antonyms: Vec<String>,
    pub last_updated: DateTime<Local>,
}

impl Vocab {
    pub fn new(
        id: String,
        word: String,
        part_of_speech: PartOfSpeech,
        definition: String,
        // examples: Vec<String>,
        // synonyms: Vec<String>,
        // antonyms: Vec<String>,
        // last_updated: DateTime<Local>,
    ) -> Self {
        Self {
            id,
            word,
            part_of_speech,
            definition,
            examples: Vec::new(),
            synonyms: Vec::new(),
            antonyms: Vec::new(),
            last_updated: Local::now(),
        }
    }
}

#[derive(Debug, Serialize)]
pub enum PartOfSpeech {
    Noun,
    Verb,
    Adjective,
    Adverb,
}
